services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,263 @@
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "User session is not found", Response.Status.BAD_REQUEST);
+        } else {
+            logger.infof(" user session id =  %s, username = %s", userSession.getId(), userSession.getUser().getUsername());
+        }
+        updateUserSessionFromClientAuth(userSession);
+
+        logger.infof("  Created User Session's id                            = %s", userSession.getId());
+        logger.infof("  Submitted in advance User Session ID Will Be Created = %s", getUserSessionIdWillBeCreated());
+
+        // authorization (consent)
+        UserConsentModel grantedConsent = session.users().getConsentByClient(realm, user.getId(), client.getId());
+        if (grantedConsent == null) {
+            grantedConsent = new UserConsentModel(client);
+            session.users().addConsent(realm, user.getId(), grantedConsent);
+            logger.info("  Consent updated : ");
+            grantedConsent.getGrantedClientScopes().forEach(i->logger.info(i.getName()));
+        }
+
+        boolean updateConsentRequired = false;
+
+        for (String clientScopeId : authSession.getClientScopes()) {
+            ClientScopeModel clientScope = KeycloakModelUtils.findClientScopeById(realm, client, clientScopeId);
+            if (clientScope != null) {
+                if (!grantedConsent.isClientScopeGranted(clientScope) && clientScope.isDisplayOnConsentScreen()) {
+                    grantedConsent.addGrantedClientScope(clientScope);
+                    updateConsentRequired = true;
+                }
+            } else {
+                logger.warnf("Client scope or client with ID '%s' not found", clientScopeId);
+            }
+        }
+
+        if (updateConsentRequired) {
+            session.users().updateConsent(realm, user.getId(), grantedConsent);
+            logger.info("  Consent granted : ");
+            grantedConsent.getGrantedClientScopes().forEach(i->logger.info(i.getName()));
+        }
+
+        event.detail(Details.CONSENT, Details.CONSENT_VALUE_CONSENT_GRANTED);
+
+        event.success();
+    }
+
+
+    protected void persistDecoupledAuthenticationResult(String status) {
+        DecoupledAuthnResult decoupledAuthnResult = new DecoupledAuthnResult(getExpiration(), status);
+        DecoupledAuthnResultParser.persistDecoupledAuthnResult(session, getAuthResultId(), decoupledAuthnResult, getExpiration());
+    }
+
+    abstract protected String getScope();
+    abstract protected String getUserSessionIdWillBeCreated();
+    abstract protected String getUserIdToBeAuthenticated();
+    abstract protected String getAuthResultId();
+    abstract protected int getExpiration();
+
+    abstract protected Response verifyDecoupledAuthnResult();
+
+    private void updateUserSessionFromClientAuth(UserSessionModel userSession) {
+        for (Map.Entry<String, String> attr : clientAuthAttributes.entrySet()) {
```

Comment:
Woudn't be easier to just `userSession.getNotes().putAll(clientAuthAttributes)`? And probably remove this method.

Reply:
@pedroigor Yes, that is right. I will write `userSession.getNotes().putAll(clientAuthAttributes)` directly on the caller of this method and remove it.